packages/application-shell-connectors/src/components/application-context/application-context.tsx

```diff
@@ -191,27 +200,47 @@ function withApplicationContext<
   };
 }
 
-// Forward-compatibility with React Hooks 🎉
-const useApplicationContext = React.useContext
-  ? <
-      SelectedContext extends {} = {},
-      AdditionalEnvironmentProperties extends {} = {}
-    >(
-      selector?: (
-        context: TApplicationContext<AdditionalEnvironmentProperties> | {}
-      ) => SelectedContext
-    ):
-      | SelectedContext
-      | TApplicationContext<AdditionalEnvironmentProperties>
-      | {} => {
-      const applicationContext = React.useContext(Context);
-      return selector ? selector(applicationContext) : applicationContext;
-    }
-  : () => {
-      throw new Error(
-        `React hooks do not seem to be available yet in the installed React version "${React.version}". Please check the React Hooks documentation for more info: https://reactjs.org/hooks.`
-      );
-    };
+// Use function overloading to declare two possible signatures with two
+// distict return types, based on the selector function argument.
+function useApplicationContext<
+  AdditionalEnvironmentProperties extends {} = {}
+>(): TApplicationContext<AdditionalEnvironmentProperties>;
+function useApplicationContext<
+  SelectedContext,
```

Comment:
I wondered if this is not the same as ReturnType typeof selector? By that consumers would not have to specify what their selector returns in many cases?

Reply:
Hmm I'm not sure I get your point. Can you clarify what you mean?

Reply:
Are you maybe referring to the conditional types, using `infer`?

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html

Reply:
Sure. What I meant is that with what we pass as a generic we state what the `selector`-fn returns? We could attempt TypeScript to infer that:

```js
function getInt(a: string) {
  return parseInt(a);
}

type A = ReturnType<typeof getInt>; // => number
```

Reply:
Yes I've been looking into that and in this case it doesn't really work because the return type of the selector function can be literally _anything_

```js
useApplicationContext(context => context.permissions)
useApplicationContext(context => Object.keys(context.permissions).length > 0)
useApplicationContext(context => "foo")
```

Which means the signature would be: 

```ts
(context: TApplicationContext) => unknown
```

which is why we need a generic type

```ts
<T>(context: TApplicationContext) => T
```

But maybe for other more deterministic cases we can see if we can use that.

Reply:
react-redux seems to do the same:

``js
export function useSelector<TState, TSelected>(
    selector: (state: TState) => TSelected,
    equalityFn?: (left: TSelected, right: TSelected) => boolean
): TSelected;
```